Lib/http/server.py

```diff
@@ -1256,39 +1253,48 @@ def test(HandlerClass=BaseHTTPRequestHandler,
 
 if __name__ == '__main__':
     import argparse
+    import contextlib
 
     parser = argparse.ArgumentParser()
     parser.add_argument('--cgi', action='store_true',
-                       help='Run as CGI Server')
+                        help='Run as CGI Server')
     parser.add_argument('--bind', '-b', metavar='ADDRESS',
                         help='Specify alternate bind address '
                              '[default: all interfaces]')
     parser.add_argument('--directory', '-d', default=os.getcwd(),
                         help='Specify alternative directory '
-                        '[default:current directory]')
+                        '[default: current directory]')
+    parser.add_argument('--protocol', '-p', default='HTTP/1.0',
```

Comment:
Have you considered default of HTTP/1.1? What does it mean to specify a protocol version? Could there be a better way to infer the supported protocol rather than require an explicit parameter? Is there a way this parameter could reflect the default of the server?

Reply:
This is the server!

Reply:
@jaraco

> Have you considered default of HTTP/1.1?

This is the HTTP version that the *server* supports. It is sent in responses to inform the client about the server’s communication capabilities (cf. [RFC 7230, § 2.6](https://datatracker.ietf.org/doc/html/rfc7230#section-2.6)). I set the default to HTTP/1.0 as this is the default value for class attribute `http.server.BaseHTTPRequestHandler.protocol_version` (cf. the [Python documentation](https://docs.python.org/3/library/http.server.html#http.server.BaseHTTPRequestHandler.protocol_version)):

> **`protocol_version`**
> 
> This specifies the HTTP protocol version used in responses. If set to 'HTTP/1.1', the server will permit HTTP persistent connections; however, your server must then include an accurate Content-Length header (using send_header()) in all of its responses to clients. For backwards compatibility, the setting defaults to 'HTTP/1.0'.

Now by backwards compatibility I am not sure what the Python documentation means. Is it referring to the following paragraph of the HTTP specification (cf. [RFC 7230, § 2.6](https://datatracker.ietf.org/doc/html/rfc7230#section-2.6))?

>   A server MAY send an HTTP/1.0 response to a request if it is known or suspected that the client incorrectly implements the HTTP specification and is incapable of correctly processing later version responses, such as when a client fails to parse the version number correctly or when an intermediary is known to blindly forward the HTTP-version even when it doesn't conform to the given minor version of the protocol.  Such protocol downgrades SHOULD NOT be performed unless triggered by specific client attributes, such as when one or more of the request header fields (e.g., User-Agent) uniquely match the values sent by a client known to be in error.

I am not an HTTP expert, so maybe @njsmith could give us his input on whether using HTTP/1.0 was and is still a sensible default.

Reply:
I don’t know if HTTP Keep-alive (the feature added by using version 1.1 here) matters that much for this tool.
`python3 -m http.server` is a nice little trick when you need to transfer files to something that has web access but not much more, it’s not a real production-grade server in any way (performance, security or features).
I think the presence of options for CGI and protocol is already plenty!

Reply:
Thanks, and if the default should change it is probably better to do it in a separate PR anyway.

By the way, I documented:

```
.. versionadded:: 3.11
     ``--protocol`` argument was introduced.
```

I chose Python 3.11 because branches 3.9 and 3.10 are in [bugfix status](https://devguide.python.org). Did I choose the correct Python version?

Reply:
@merwok By the way, I documented:

.. versionadded:: 3.11
     ``--protocol`` argument was introduced.

I chose Python 3.11 because branches 3.9 and 3.10 are in bugfix status. Did I choose the right Python version?

Reply:
The fact that this PR does two things now makes backporting hard. I feel like the new `--protocol` version is a new feature and should not be backported, but the `-d` fix (where CGI ignored it) is a bugfix and should be backported.

Reply:
Agreed!  Sorry for not seeing that earlier, we have two tickets and it’s easier to have two PRs.  Do you mind separating them?

Reply:
Good idea, done! We now have two PRs:

1. [PR 30701](https://github.com/python/cpython/pull/30701) (this PR) which fixes `-d`/`--directory` when used with `--cgi` (associated with [issue 46436](https://bugs.python.org/issue46436));
2. [PR 30999](https://github.com/python/cpython/pull/30999) which adds `-p`/`--protocol` (associated with [issue 46285](https://bugs.python.org/issue46285)).